### ISSUE TITLE
EPIC: Tier-Filtered Model Retraining (v2.1 Classification-Based)

### ISSUE THREAD

**Issue by marctjones** [marctjones/klareco#616]
# Overview

Retrain learned models using tier-filtered vocabulary from v2.1 database classification AND establish standard AST annotation interface.

## Core Architectural Principle

**Klareco maximizes deterministic processing.** Models should:
- ✅ **Consume ASTs** (not raw text) - grammar/structure already parsed
- ✅ **Add annotations** (not re-learn deterministic features)
- ✅ **Use standard interface** (ASTAnnotator protocol)
- ❌ **Never re-parse** what M0 already extracted
- ❌ **Never learn grammar** (case, tense, gender, number) - it's in the AST!

## Problem

1. **Current root embeddings** (10,819 vocab) include **function words** causing **embedding collapse**
2. **No standard interface** for models to consume/annotate ASTs
3. **Models may re-learn** what AST already provides deterministically

### Part A: Tier-Filtered Vocabulary (v2.1 Classification)

Retrain models with clean vocabularies excluding:
- **Tier 0** (187 grammatical words): Handled deterministically by M0 parser
- **Tier 5** (1M+ parse failures): Garbage data

Include only **content words**:
- **Tier 1a**: 750 Unua Libro roots
- **Tier 1b**: 1,403 Fundamento roots
- **Tier 2**: 7,646 ReVo roots
- **Total**: ~9,800 clean content word roots

### Part B: Standard AST Annotation Interface

All models follow protocol:
```python
class ASTAnnotator:
    def annotate(self, ast: dict) -> dict:
        """Add annotations to AST, never re-compute deterministic features."""
```

Each model layer:
1. Reads AST structure (from M0 parser)
2. Reads previous annotations (from earlier layers)
3. Adds its own annotations
4. Returns enriched AST

**No model trains on grammar features** - they're deterministic!

## Model Taxonomy: Deterministic vs Learned

| Model | Deterministic | Learned | What Learning Adds |
|-------|---------------|---------|-------------------|
| **M0 Parser** | 100% | 0% | Nothing (16 grammar rules) |
| **Stage 1 Roots** | 0% | 100% | Semantic similarity of content words |
| **Compositional** | 50% | 50% | Endings deterministic, affix transforms learned |
| **M1 Selectional** | 0% | 100% | Semantic plausibility from corpus |
| **Entity Classifier** | 70% | 30% | Tier 1-2 deterministic, Tier 3 learned |
| **M2.1 Taxonomy** | **90%** | **10%** | ReVo+ConceptNet+affixes deterministic, OOV learned |
| **M2.2 Coreference** | **80%** | **20%** | Gender/number/case deterministic, ambiguity learned |
| **Reranker** | 20% | 80% | Structural filtering deterministic, relevance learned |

### Phase 0: AST Annotation Infrastructure (NEW)
- [ ] #633: Design standard ASTAnnotator interface
- [ ] #634: Implement AST annotation protocol in base classes
- [ ] #635: Update M0 parser to output standard annotated AST format
- [ ] #636: Create AST annotation validation tests

### Phase 1: Root Embeddings (CRITICAL - Blocking)
- [ ] #617: Generate tier-filtered training data from v2.1 DB
- [ ] #618: Train root embeddings (~9,800 vocab, exclude tier0)
- [ ] #619: Test quality (no collapse, semantic clusters)
- [ ] #620: Integration (freeze for downstream, implement ASTAnnotator interface)

### Phase 2: M1 Selectional Preference
- [ ] #621: Generate (subject, verb, object) training data from v2.1 DB
- [ ] #622: Train M1 with frozen root embeddings + ASTAnnotator interface
- [ ] #623: Test selectional accuracy (target: 80%+)
- [ ] #624: Integration into RAG query expansion

### Phase 3: Entity Classifier (Mostly Deterministic!)
- [ ] #625: Generate semantic annotation training data from v2.1 DB
- [ ] #626: Implement deterministic Tier 1-2 classification (affixes, vortspeco)
- [ ] #627: Train Tier 3 fine-grained classifier (~5M params for 10% of cases)
- [ ] #628: Integration into RAG entity boosting

### Phase 4: Reranker
- [ ] #629: Generate query-document relevance pairs from v2.1 DB
- [ ] #630: Train reranker with frozen root embeddings + ASTAnnotator interface
- [ ] #631: Test relevance ranking (MRR > 0.6)
- [ ] #632: Integration into production RAG pipeline

### Phase 5: M2.1 Taxonomy (Mostly Deterministic - NEW)
- [ ] #637: Load ReVo IS-A relations into graph
- [ ] #638: Load ConceptNet taxonomy relations
- [ ] #639: Implement deterministic affix taxonomy rules (90% coverage)
- [ ] #640: Train small fallback model for OOV words (10% of cases)
- [ ] #641: Integration into semantic reasoning pipeline

### Phase 6: M2.2 Coreference (Mostly Deterministic - NEW)
- [ ] #642: Implement deterministic pronoun resolution (gender/number/case)
- [ ] #643: Implement recency heuristics for unambiguous cases (80% coverage)
- [ ] #644: Train small disambiguation model for hard cases (20% of cases)
- [ ] #645: Integration into multi-sentence understanding

### Technical Validation
- [ ] Root embeddings: No function words in vocabulary
- [ ] Root embeddings: Mean pairwise similarity < 0.5 (no collapse)
- [ ] Root embeddings: Semantic clusters separated (gap > 0.03)
- [ ] M1 selectional: Accuracy ≥ 80% on test set
- [ ] Entity classifier: Tier 1-2 accuracy ≥ 95% (deterministic baseline)
- [ ] Entity classifier: Tier 3 accuracy ≥ 70% (learned)
- [ ] Reranker: MRR ≥ 0.6
- [ ] M2.1 Taxonomy: 90%+ coverage from deterministic rules
- [ ] M2.2 Coreference: 80%+ accuracy from deterministic rules
- [ ] All models use standard ASTAnnotator interface

### Integration Validation
- [ ] All models integrated into working RAG pipeline
- [ ] End-to-end RAG latency < 2 seconds
- [ ] AST passes through all layers with cumulative annotations
- [ ] No model re-learns deterministic features

## Learned Parameters Inventory

After Epic #616:
- Stage 1 Root Embeddings: 320K
- Compositional Affixes: 180K
- M1 Selectional: 10M
- Entity Classifier (Tier 3): ~1.5M (not 5M, only for 10% hard cases)
- Reranker: 180K
- M2.1 Taxonomy (OOV fallback): ~1M (small)
- M2.2 Coreference (disambiguation): ~2M (small)
- **Total: ~15M params** (down from originally estimated 15.5M)

## Dependencies

**Sequential execution**:
1. Phase 0 (AST interface) + Phase 1 (Root Embeddings) - **must complete first**
2. Phases 2-6 can proceed in parallel after Phase 0-1 complete
3. Final integration requires all phases complete

## References

- v2.1 Classification: docs/V2.1_DATABASE_CLASSIFICATION_COMPLETE.md
- Architecture: VISION.md (decomposable contributions principle)
- Training readiness: docs/TRAINING_READY_SUMMARY.md
- Validation: scripts/validate_tier_filtered_training.py
- Tier-filtered vocab: data/vocabularies/tier_filtered_roots.json

**Comment by marctjones** [comment]
## CLI Infrastructure Update

Added comprehensive CLI architecture to support model development and debugging:

### New Epics Created

- **Epic #637**: CLI Phase 1 - Inspection Tools (CRITICAL for development)
  - `klareco inspect ast` - Inspect AST structure
  - `klareco inspect annotations` - Show all annotations
  - `klareco inspect tensor` - Decode embeddings to similar words
  - `klareco annotate text` - Run annotation pipeline

- **Epic #640**: CLI Phase 2 - Training Lifecycle
  - `klareco train roots` - Train root embeddings
  - `klareco train m1` - Train M1 selectional
  - Unified training interface, consistent checkpointing

- **Epic #641**: CLI Phase 3 - Data Pipeline
  - `klareco data export roots` - Export training data from v2.1 DB
  - Database as single source of truth
  - On-demand training data generation

- **Epic #642**: CLI Phase 4 - Pipeline Debugging
  - `klareco inspect pipeline` - Trace annotation flow
  - `klareco query explain` - Explain retrieval decisions
  - Full explainability support

### Impact on Epic #616

The CLI infrastructure supports the model retraining workflow:

1. **Data Export** (Epic #641): Generate training data from v2.1 DB
2. **Training** (Epic #640): Train models via CLI
3. **Inspection** (Epic #637): Debug models during development
4. **Testing** (Epic #642): Validate model quality

### Documentation

Complete design: `docs/CLI_ARCHITECTURE.md`
- Naming conventions for scripts
- Version compatibility matrix
- Docstring standards
- Coherent command structure

### Priority

Epic #637 (Inspection Tools) is **CRITICAL** - blocks all Phase 0-1 development in Epic #616.
Must implement inspection tools before starting Phase 1 (Root Embeddings).
